teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,479 @@
+      [`git-clone` Task update in the Catalog][catalog-pr-711], the versions of those
+      `Tasks` have to be bumped with this update.
+    - Catalog `Tasks` that leverage this pattern to intentionally inject executable code
+      (e.g. the `git-cli` Task [accepts a script to run as a param][git-cli]) can be updated
+      to execute the contents of the `Parameter` from a file or call `eval` on it explicitly.
+
+  1. Tekton Pipelines and Catalog documentation should be updated to the new approach.
+
+1. **`Tasks` should not support implicit runtime code injection**.
+
+   Disallow interpolating `Parameter` variables directly into the body of a `Script`
+   by *Tekton Pipelines V1* release to be secure by default.
+
+
+### Use Case
+
+As described in the [original proposal][pipeline-issue-781], the use case for `Script`
+in `Tasks` is:
+
+> Users need an easy cruft-free way to express a multi-statement `Script` in the
+> body of a `Task` `Step`, without having to understand advanced topics about
+> `Containers`, like what an `entrypoint` or a `command` is.
+>
+> Users should only need to have passing familiarity with a shell environment
+> to be successful using Tekton.
+
+Problems with variable subsitution was noted in [original proposal][pipeline-issue-781]:
+
+> How this syntax works with input substitutions, e.g., ``${inputs.params.flags}`` might
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+1. **[Secure by default][secure-by-default]**: Out of the box
+   *Tekton Pipelines* should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards compatible in v1beta1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1beta1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **Backwards incompatible in v1**. Starting with v1, Tekton should strive to
+   be "secure by default" out of the box and so inclusion of param variables
+   in `script` fields should be treated as an error. Newly submitted (or resolved)
+   `Tasks` with this error should fail validation. Existing `Tasks` that include this
+   error should be prevented from executing.
+1. **Provide an escape hatch in v1**. If an operator decides that they want to
```

Comment:
When you say "operator" do you mean the operator of a cluster has the choice of whether to permit interpolation? Do you know why users might want this? (Maybe they are trying to use a task they don't own that has params embedded in a script-- but ideally as you mentioned we should be able to automate updating Tasks, and all catalog tasks should be updated.)

I like that allowing this behavior to be configurable by cluster operators doesn't force everyone to audit catalog Tasks for security, but I'm not sure why someone might need it.

Reply:
Yeah, that's the definition of "operator" we had in mind.

The intention was to support users who have legacy Tasks that haven't been updated to access params in whatever way we decide here. Automation should be able to fix 90% of cases I hope and validation could disallow the rest, but the idea was to give operators that extra wiggle room if they really really don't have the bandwidth to update an existing Task that's vital to some part of their infra.

I definitely don't have any user reports to hand requesting this kind of support though.